bindings/matrix-sdk-ffi/src/client_builder.rs

```diff
@@ -238,6 +378,109 @@ impl ClientBuilder {
     pub async fn build(self: Arc<Self>) -> Result<Arc<Client>, ClientBuildError> {
         Ok(Arc::new(self.build_inner().await?))
     }
+
+    /// Finish the building of the client and attempt to log in using the
+    /// provided [`QrCodeData`].
+    ///
+    /// This method will build the client and immediately attempt to log the
+    /// client in using the provided [`QrCodeData`] using the login
+    /// mechanism described in [MSC4108]. As such this methods requires OIDC
+    /// support as well as sliding sync support.
+    ///
+    /// The usage of the progress_listener is required to transfer the
+    /// [`CheckCode`] to the existing client.
+    ///
+    /// [MSC4108]: https://github.com/matrix-org/matrix-spec-proposals/pull/4108
+    pub async fn build_with_qr_code(
+        self: Arc<Self>,
+        qr_code_data: &QrCodeData,
+        oidc_configuration: &OidcConfiguration,
+        progress_listener: Box<dyn QrLoginProgressListener>,
+    ) -> Result<Arc<Client>, HumanQrLoginError> {
+        if let QrCodeModeData::Reciprocate { homeserver_url } = &qr_code_data.inner.mode_data {
+            let mut builder = self.server_name_or_homeserver_url(homeserver_url.to_string());
+            let uuid = uuid::Uuid::new_v4().to_string();
+
+            // If a base directory was configured, create a random subdirectory, we will
+            // rename this directory later on.
+            let directories = if let Some(base_path) = &builder.base_path {
+                let base_directory = PathBuf::from(base_path);
+                let random_dir = base_directory.join(&uuid);
+
+                tracing::debug!("Setting the Client storage path to {random_dir:?}");
+
+                builder = builder.base_path(
+                    random_dir
+                        .to_str()
+                        .expect("The base path and the uuid both are valid UTF-8 strings")
+                        .to_string(),
+                );
+
+                Some((base_directory, random_dir))
+            } else {
+                None
+            };
+
+            let client = builder.build().await.map_err(|e| {
+                tracing::error!("Couldn't build the client {e:?}");
+                HumanQrLoginError::Unknown
+            })?;
+
+            if client.sliding_sync_proxy().is_none() {
+                return Err(HumanQrLoginError::SlidingSyncNotAvailable);
+            }
+
+            let client_metadata = oidc_configuration
+                .try_into()
+                .map_err(|_| HumanQrLoginError::OidcMetadataInvalid)?;
+
+            let oidc = client.inner.oidc();
+            let login = oidc.login_with_qr_code(&qr_code_data.inner, client_metadata);
+
+            let mut progress = login.subscribe_to_progress();
+
+            let _progress_task = TaskHandle::new(RUNTIME.spawn(async move {
```

Comment:
a `TaskHandle::drop` will abort the task, so surely this `_progress_task` must either be returned to the caller, or stashed somewhere else?

Reply:
Why? No further progress will be made once the function completes.

Reply:
Oh, I missed that the following code could fail thereafter, so it's just a small API crime in that we're abusing a public API to abort on drop, correct? If so, I *think* we had some `abort_on_drop()` extension method on a trait that achieves the same. But at this point, it's ok to keep `TaskHandle`, maybe add a comment explaining why it's fine to not use the task handle, please? since another reader may go through the same steps as I did (I know future-me would surely appreciate this).

Reply:
I added a comment, though is it an API crime to use a public API? We just want a task that will not detach itself implicitly. Other executors might create tasks which require explicit detachment and the default is the cancellation.

I don't think that this is terribly important since the stream should end, once the future is dropped the `SharedObservable` is dropped, but the cancellation shouldn't hurt either as a precaution.